lib/srv/discovery/access_graph.go

```diff
@@ -0,0 +1,304 @@
+func (s *Server) getAllAWSSyncFetchers() []aws_sync.AWSSync {
+	allFetchers := make([]aws_sync.AWSSync, 0, len(s.dynamicTAGSyncFetchers))
+
+	s.muDynamicTAGSyncFetchers.RLock()
+	for _, fetcherSet := range s.dynamicTAGSyncFetchers {
+		allFetchers = append(allFetchers, fetcherSet...)
+	}
+	s.muDynamicTAGSyncFetchers.RUnlock()
+
+	allFetchers = append(allFetchers, s.staticTAGSyncFetchers...)
+	// TODO: submit fetchers event
+	return allFetchers
+}
+
+func pushUpsertInBatches(
+	client accessgraphv1alpha.AccessGraphService_AWSEventsStreamClient,
+	upsert *accessgraphv1alpha.AWSResourceList,
+) error {
+	for i := 0; i < len(upsert.Resources); i += batchSize {
+		end := i + batchSize
+		if end > len(upsert.Resources) {
+			end = len(upsert.Resources)
+		}
+		err := client.Send(
+			&accessgraphv1alpha.AWSEventsStreamRequest{
+				Operation: &accessgraphv1alpha.AWSEventsStreamRequest_Upsert{
+					Upsert: &accessgraphv1alpha.AWSResourceList{
+						Resources: upsert.Resources[i:end],
+					},
+				},
+			},
+		)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+func pushDeleteInBatches(
+	client accessgraphv1alpha.AccessGraphService_AWSEventsStreamClient,
+	delete *accessgraphv1alpha.AWSResourceList,
+) error {
+	for i := 0; i < len(delete.Resources); i += batchSize {
+		end := i + batchSize
+		if end > len(delete.Resources) {
+			end = len(delete.Resources)
+		}
+		err := client.Send(
+			&accessgraphv1alpha.AWSEventsStreamRequest{
+				Operation: &accessgraphv1alpha.AWSEventsStreamRequest_Delete{
+					Delete: &accessgraphv1alpha.AWSResourceList{
+						Resources: delete.Resources[i:end],
+					},
+				},
+			},
+		)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+func push(
+	client accessgraphv1alpha.AccessGraphService_AWSEventsStreamClient,
+	upsert *accessgraphv1alpha.AWSResourceList,
+	delete *accessgraphv1alpha.AWSResourceList,
+) error {
+	err := pushUpsertInBatches(client, upsert)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = pushDeleteInBatches(client, delete)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = client.Send(
+		&accessgraphv1alpha.AWSEventsStreamRequest{
+			Operation: &accessgraphv1alpha.AWSEventsStreamRequest_Sync{},
+		},
+	)
+	return trace.Wrap(err)
+}
+
+// NewAccessGraphClient returns a new access graph service client.
+func newAccessGraphClient(ctx context.Context, certs []tls.Certificate, config servicecfg.AccessGraphConfig, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	opt, err := grpcCredentials(config, certs)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	conn, err := grpc.DialContext(ctx, config.Addr, append(opts, opt)...)
+	return conn, trace.Wrap(err)
+}
+
+// initializeAndWatchAccessGraph initializes the access graph service and watches the auth server for events.
+// This function acquires a lock on the backend to ensure that only one instance of auth server is sending
+// events to the access graph service at a time.
```

Comment:
I don't think this part of the comment applies

Reply:
Yeah. This comment belongs to a subsequent PR. I removed it